source/retryable-reads/tests/unified/handshakeError.yml

```diff
@@ -0,0 +1,68 @@
+description: "retryable reads handshake failures"
+
+schemaVersion: "1.0"
+
+runOnRequirements:
+  - minServerVersion: "4.2"
+    topologies: [replicaset, sharded, load-balanced]
+    auth: true
+
+createEntities:
+  - client:
+      id: &client0 client0
+      useMultipleMongoses: false
+      observeEvents: [commandStartedEvent, connectionCheckOutStartedEvent]
+      uriOptions: { retryReads: true, retryWrites: true, maxPoolSize: 5 }
+  - database:
+      id: &database0 database0
+      client: *client0
+      databaseName: &databaseName retryable-handshake-tests
+  - collection:
+      id: &collection0 collection0
+      database: *database0
+      collectionName: &collectionName coll
+
+initialData:
+  - collectionName: *collectionName
+    databaseName: *databaseName
+    documents:
+      - { _id: 1, x: 11 }
+      - { _id: 2, x: 22 }
+      - { _id: 3, x: 33 }
+
+tests:
+  - description: "InsertOne succeeds after retryable handshake network error"
+    operations:
+      - name: failPoint # fail the next connection establishment
+        object: testRunner
+        arguments:
+          client: *client0
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 2 }
+            data:
+              failCommands: [saslContinue, dropCollection] # use saslContinue here to avoid SDAM errors
+              closeConnection: true
+      - name: dropCollection
+        object: *database0
+        expectError:
+          isError: true
+      - name: findOne
+        object: *collection0
+        arguments:
+          filter: { _id: 2 }
+        expectResult: { _id: 2, x: 22 }
+
+    expectEvents:
+      - client: *client0
+        eventType: cmap
+        events:
+          - { connectionCheckOutStartedEvent: {} }
+          - { connectionCheckOutStartedEvent: {} }
```

Comment:
Does this entire spec change only apply to drivers with connection pools? I'm reading through [DRIVERS-746](https://jira.mongodb.org/browse/DRIVERS-746) and although it does say:

> * there is no idle connection already available in the pool
> * the driver attempts to open a new connection to the server and complete the connection handshake

...I don't think it technically requires a pool. The main issue has to do with a retryable operation using an entirely new connection, which would seem applicable to _all_ drivers (even single-threaded libmongoc w/o a pool).

If I'm not mistaken, is there a more portable way we can write these tests? Perhaps setting the fail point with one client and then using a separate client to run the retryable read/write as its first operation? And if we go that route, would there still be a need to require auth and fail on `saslContinue` instead of a `hello` command?

Reply:
Sorry I missed this comment in my initial responses.

That sounds like it may work, but if we set the failPoint to hello, wouldn't serverSelection fail? monitoring won't be able to identify the topology, or at least it will clear the first failpoint if we use the "times: 1" method. And it wouldn't make it to the retryablity section of executing an operation. Maybe there's a different order of operations I'm not considering?

Reply:
> but if we set the failPoint to hello, wouldn't serverSelection fail?

That makes sense. Forget I suggested that. I was recently using `hello` failpoints (albeit for SDAM monitoring tests), which is probably why that was on my mind.

I reckon that failing on `hello` would actually be much worse for single-threaded clients due to [`cooldownMS`](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-monitoring.rst#cooldownms) blocking for five seconds. I don't think a failure on `saslContinue` would trigger that.

Apologies for wasting your time here. I probably have no business reviewing anything with CMAP in it.

Reply:
Grasping the tricks we're using here to test took me a min since it isn't very direct, so fresh eyes/perspective is always welcome!